components/panel/body.js

```diff
@@ -36,6 +36,14 @@ class PanelBody extends Component {
 		}
 	}
 
+	componentWillReceiveProps( nextProps ) {
+		if ( this.props.initialOpen !== nextProps.initialOpen ) {
```

Comment:
like its name suggest `intialOpen` should not change and if it does we should console.error. It means this panel is "uncontrolled". if we want a "controlled" panel body we can use the "opened" prop.

Can I aks why did you need this change?

Reply:
When changing blocks for example from a button to a paragraph. New props are passed, a new component is not built which makes sense so if you collapsed it in a block you can keep it collapsed when changing it. But if changing to different initial open settings without this the state keeps the same and the setting is useless.

Reply:
> a new component is not built which

I wonder why it's not the case, the block inspector of a button and a paragraph are complete separate components, I expect them to "remount" when switching between these two.

If they're not, I suppose this could create issues elsewhere. (any uncontrolled state value)

Reply:
I also expected them to remount but unfortunately, that is not the case, they just receive new props. Maybe there is an improvement that can make them remount but sure how.

Reply:
You can assign a unique `key` prop to trigger remounts.

Reply:
This is a really weird bug. I suspect some relation with slot-and-fill @aduth

Reply:
>This is a really weird bug. I suspect some relation with slot-and-fill @aduth

I don't know, it seems like React may just be recycling the same element.

Reply:
I expect if you give a unique key to `InspectorControls` here:

https://github.com/WordPress/gutenberg/blob/aee35468ebc4159a134672869cc6d69162e30d6a/blocks/library/paragraph/index.js#L114

Like `paragraph-inspector`... that it may resolve.

...but this sounds like a tricky implementation detail we shouldn't be expecting of block authors.

Reply:
Shouldn't we assume that `slot-fill` ensures the remounting?

Reply:
If it's a different fill, it should remount

Reply:
Hi @aduth, @youknowriad, I tried the key fix in inspector controls and I also tried adding keys to PanelColor but the components continue to be reused. It's really strange.

Reply:
It looks like react tries to reuse the components whatever it cans and although keys are a hint there is no guarantee a component will not be reused.

Reply:
It looks like the correct way is to manage state from the block components.